### PR TITLE
Adds SAVEDDATA plugin output support (text after ||) and SERVICESAVEDDATA and HOSTSAVEDDATA macros

### PR DESCRIPTION
**Pull request by willixix**
As previously discussed on nagios-devel mail list, this patch modifies interpretation of nagios plugin output putting everything after || (which can follow status output directly or follow performance data separated from it by single |) into special data field of service and host objects, which thereafter are available to plugins as $SERVICESAVEDDATA$ and $HOSTSAVEDDATA$ macros.

I was unable to find your most recent development branch code with I/O broker (wproc is over 6 months old...) so this is a patch against 3.41 nagios master code. As this modifies at least one neb function and changes plugin data results interpretation in major way, it is not suitable for 3.x. The code itself is mainly based on how perf_data is dealt with. parse_check_output() function is more than half re-written without use of unsafe strtok (I'd like to remove it from all nagios code actually). I tested the code with development version of check_netint plugin (https://github.com/willixix/WL-NagiosPlugins/blob/master/check_netint.pl) but my plugins only return single line output so this needs more testing with multi-line output plugins. Also documentation for this will need to be written, which I'll do once its clear what 4.0 nagios branch is.

### CODE REVIEW

**Comment by ageric** [comment]
On 08/15/2012 05:08 AM, William Leibzon wrote:

> As previously discussed on nagios-devel mail list, this patch
> modifies interpretation of nagios plugin output putting everything
> after || (which can follow status output directly or follow
> performance data separated from it by single |) into special data
> field of service and host objects, which thereafter are available to
> plugins as $SERVICESAVEDDATA$ and $HOSTSAVEDDATA$ macros.
> 
> I was unable to find your most recent development branch code with
> I/O broker (wproc is over 6 months old...)

It's in the Nagios trunk now-adays.

> so this is a patch against
> 3.41 nagios master code. As this modifies at least one neb function
> and changes plugin data results interpretation in major way, it is
> not suitable for 3.x. The code itself is mainly based on how
> perf_data is dealt with. parse_check_output() function is more than
> half re-written without use of unsafe strtok (I'd like to remove it
> from all nagios code actually). I tested the code with development
> version of check_netint plugin
> (https://github.com/willixix/WL-NagiosPlugins/blob/master/check_netint.pl)
> but my plugins only return single line output so this needs more
> testing with multi-line output plugins. Also documentation for this
> will need to be written, which I'll do once its clear what 4.0 nagios
> branch is.

I really like the way this is going, but unfortunately, the patch is
completely littered with unrelated nonsensical changes, like this one:
@@ -1151,5 +1151,5 @@ int read_main_config_file(char *main_config_file) {
            low_service_flap_threshold = strtod(value, NULL);
            if(low_service_flap_threshold <= 0.0 || low_service_flap_threshold >= 100.0) {
-               asprintf(&error_message, "Illegal value for low_service_flap_threshold");
-               (void)asprintf(&error_message, "Illegal value for low_service_flap_threshold");
              error = TRUE;

Care to re-roll against the Nagios trunk and without the cruft?
Also, the 'master' branch of my git repo contains the latest
worker process code as of about 5 minutes ago, so you can build
on top of that if you prefer to use git.

**Comment by willixix** [comment]
I changed configure.in to add "-Wstrict-prototypes -Wno-unused-result " which gets rid of asprintf warnings and I'm keeping it (but you should rerun autoconf) so I'll remove voids. Your latest build merged fine with my patch, I've already got it running on my backup nagios server.

BTW, I'm actually working on more "nonsensical" changes in a separate branch. Mostly trying to get rid of all strtok everywhere in the code and any other function I notice that is not thread safe. After I'll actually try to add multi-threaded support despite how everyone is sayings its not possible. This is going to separate branch so you don't have to quite decide about any of it right now. If I get it to work, I'll talk to both you and Ethan.

**Comment by ageric** [comment]
On 08/15/2012 03:47 PM, William Leibzon wrote:

> I changed configure.in to add "-Wstrict-prototypes -Wno-unused-result
> " which gets rid of asprintf warnings and I'm keeping it (but you
> should rerun autoconf) so I'll remove voids.

I have no idea how you got those warnings in the first place. I don't.
What system are you on?

> Your latest build merged
> fine with my patch, I've already got it running on my backup nagios
> server.

Ah, great. Then it should be possible to rebase it onto the updated
master branch. I still have to work with svn when I'm fiddling with
Nagios, so merges are a no-go I'm afraid.

> BTW, I'm actually working on more "nonsensical" changes in a separate
> branch. Mostly trying to get rid of all strtok everywhere in the code
> and any other function I notice that is not thread safe.

Getting rid of -Wno-unused-result warnings by casting it to (void) is
nonsensical. Getting rid of strtok() is sensible but sort of unnecessary.
Nagios isn't threaded anymore.

Btw, I wasn't arguing against casting ignored return values to void
per se, but I don't want to see hundreds of such lines in a patch that
does something else too. There are plenty of janitor commits in Nagios
already, but they all do _only_ janitorial work.

> After I'll
> actually try to add multi-threaded support despite how everyone is
> sayings its not possible. This is going to separate branch so you
> don't have to quite decide about any of it right now. If I get it to
> work, I'll talk to both you and Ethan.

Multithreaded support for what? There's really nothing in Nagios that
needs threads anymore, and fork() is 10 times faster on a small process
than it is on a large. Adding threads to the mix will only make fork()
even slower, and that's what's limiting us right now. Building support
for specialized workers so they can take over checks for certain configured
commands and run several such checks _without_ forking would be immensely
useful though, but it's your time so you spend it anyway you want.

A word of advice though; If you notice that Nagios with threads doesn't
outperform Nagios without threads by at least 20%, you may as well drop
it early and spend your time on something more profitable, because the
added complexity that threads incur everywhere just isn't worth it for
anything less than substantial gains.

**Comment by willixix** [comment]
Your latest commits (I haven't pulled for 2 weeks until today) contained a fatal error. In objects.c in free_objects_data() function a loop freeing contact addresses is using "i" as iterator and outer loop for contacts is using it too. Looking for it took me all day and while searching for what's causing memory errors, I went through entire objects.c cleaning its buggy error handling on memory allocation. Plus also few other, essentially esthetics, changes as I hate when ++ and -- prefix and suffix are used in assignment. I'm afraid those changes got added to my pull request as git just does it automatically.

On our previous thread:

I'm using latest Ubuntu 12.04 on dev server and Ubuntu 11.10 on my laptop. In both cases I had a slew of errors for all functions with unused return values. Very tiresome. But similar behavior should actually happen on all systems with new gcc 4.6 and above. BTW, Ubuntu 11.10 comes with old autoconf which does not like AC_HELP_STRING (they did add new one in updates though). And no, I'm not big fan of "non-essential" janitorial work either, I just thought at first I'd be able to deal with it by casting to void (which is essentially correct way to do it) but with so many, I gave up at some point and just changed configure.

I did get rid of all strtok in nagios (except CGIs) in my development branch. So if you want it will be possible to merge this later as well. I have not worked on it more yet and trying to introduce threads would require quite a bit of work in addition to just strtok. Its probably 2 weeks to month worth of work and I simply don't know if I have the time left before quarter begins to work on it, especially since most of September I'll be doing consulting work on puppet and not have much time for essentially unpaid nagios development work.

Multi-threaded support could be used to process/rip results from checks, schedule them, run brokers, etc. If its done properly, it'd allow nagios to scale to millions of services (ok, maybe not that many, but you get the point). What you did is to replace old nagios system of running checks with much better broker, but that is basically what dnx and mod_gearman were doing already by moving the actual checks to external process and servers. Also threads are in fact always faster than processes, for Linux see http://www.kernel.org/doc/man-pages/online/pages/man2/clone.2.html on how much difference there is between what it is done when creating a thread and what fork does (with all options set it would be fork, but standard thread model is not like that); that said running plugins, which are all separate scripts and programs, is probably better/safer with multi-process broker model like you are doing because you have to use system and exec anyway.

**Comment by ageric** [comment]
On 08/29/2012 11:13 AM, William Leibzon wrote:

> Your latest commits (I haven't pulled for 2 weeks until today)
> contained a fatal error. In objects.c in free_objects_data() function
> a loop freeing contact addresses is using "i" as iterator and outer
> loop for contacts is using it too.

That was fixed a few days ago. Thanks for noticing though.

> Looking for it took me all day and
> while searching for what's causing memory errors, I went through
> entire objects.c cleaning its buggy error handling on memory
> allocation.

That would be nice to have in a separate patch.

> Plus also few other, essentially esthetics, changes as I
> hate when ++ and -- prefix and suffix are used in assignment.

_sigh_ Please don't make pretty while making changes. Especially when
"pretty" is subject to your opinion and you just _know_ I won't take
the patch if it's got a billion unrelated things intermingled.

If you happen to get a typo in there turning + to - we'd be totally
screwed when trying to review a humongous patch for bugs.

> I'm
> afraid those changes got added to my pull request as git just does it
> automatically.

Are they at least in a separate patch?

> On our previous thread:
> 
> I'm using latest Ubuntu 12.04 on dev server and Ubuntu 11.10 on my
> laptop. In both cases I had a slew of errors for all functions with
> unused return values. Very tiresome. But similar behavior should
> actually happen on all systems with new gcc 4.6 and above. BTW,
> Ubuntu 11.10 comes with old autoconf which does not like
> AC_HELP_STRING (they did add new one in updates though). And no, I'm
> not big fan of "non-essential" janitorial work either, I just thought
> at first I'd be able to deal with it by casting to void (which is
> essentially correct way to do it) but with so many, I gave up at some
> point and just changed configure.
> 
> I did get rid of all strtok in nagios (except CGIs) in my development
> branch. So if you want it will be possible to merge this later as
> well. I have not worked on it more yet and trying to introduce
> threads would require quite a bit of work in addition to just strtok.
> Its probably 2 weeks to month worth of work and I simply don't know
> if I have the time left before quarter begins to work on it,
> especially since most of September I'll be doing consulting work on
> puppet and not have much time for essentially unpaid nagios
> development work.

Getting rid of strtok() is indeed nice.

> Multi-threaded support could be used to process/rip results from
> checks, schedule them, run brokers, etc. If its done properly, it'd
> allow nagios to scale to millions of services (ok, maybe not that
> many, but you get the point). What you did is to replace old nagios
> system of running checks with much better broker, but that is
> basically what dnx and mod_gearman were doing already by moving the
> actual checks to external process and servers.

Yes, but not everyone were using them, and they required additional
hardware that were outside of Nagios' control and basically slowing
development down rather than making it speed ahead forward. The idea
behind modules was always to
a) Implement features that were outside the scope of the core
b) Experiment with new designs that may later be implemented into
   core Nagios

dnx, mod_gearman and Merlin all fall into 'b' (with merlin being
sided halfway between a and b since it also does database inserts),
and livestatus falls solely in a, but is creeping over towards b
as Mathias has added his "livecheck" thing. Modules that fall in b
should, if successful, get built in to Nagios, and that's what's
happening now.

> Also threads are in
> fact always faster than processes, for Linux see
> http://www.kernel.org/doc/man-pages/online/pages/man2/clone.2.html on
> how much difference there is between what it is done when creating a
> thread and what fork does (with all options set it would be fork, but
> standard thread model is not like that); that said running plugins,
> which are all separate scripts and programs, is probably better/safer
> with multi-process broker model like you are doing because you have
> to use system and exec anyway.

And shared memory is always faster than socket-based IPC. However, it
lacks the nice process separation one gets for free with socket-based
IPC, not to mention reuse of code intended for future work to get
cross-server communication.

> --- Reply to this email directly or view it on GitHub: 
> https://github.com/ageric/nagios/pull/1#issuecomment-8119472

Thanks. I'd rather not. I have no idea if you even get this email, but
making it harder for me to reply to your email is not a good way to
communicate.

**Comment by ageric** [comment]
@@ -1062,4 +1063,10 @@ contact _add_contact(char *name, char *alias, char *email, char *pager, char *_a
    new_contact->service_notification_period_ptr = stp;
 #endif
-   new_contact->custom_variables = NULL;
-   new_contact->name = NULL;
-   new_contact->alias = NULL;
-   new_contact->email = NULL;
-   new_contact->pager = NULL;
  +
  if((new_contact->name = (char *)strdup(name)) == NULL)
      result = ERROR;

This is crap, since contact_table is calloc()'ed every single byte in it is set to zero, so all pointers are already NULL. Same with all the other similar chunks.

@@ -1132,16 +1142,25 @@ contact _add_contact(char *name, char *alias, char *email, char *pager, char *_a
    if(result == ERROR) {
        for(x = 0; x < MAX_CONTACT_ADDRESSES; x++)
-           my_free(new_contact->address[x]);
-       my_free(new_contact->name);
-       my_free(new_contact->alias);
-       my_free(new_contact->email);
-       my_free(new_contact->pager);
-           if (new_contact->address[x] != NULL)
-               my_free(new_contact->address[x]);
-       if(new_contact->name!=NULL)
-           my_free(new_contact->name);
-       if(new_contact->alias!=NULL)
-           my_free(new_contact->alias);
-       if(new_contact->email!=NULL)
-           my_free(new_contact->email);
-       if(new_contact->pager!=NULL)
-           my_free(new_contact->pager);

This is crap, since my_free() is a macro that explicitly checks for NULL-ness before free()ing. Same with other similar chunks. These two types of changes tells me you've browsed the code and changed it in ways that looks as if they make sense but you haven't really understood what it does or the components involved. Naturally, that makes me a tad sceptical towards anything else you might throw at me.
-   new_contact->id = num_objects.contacts++;
-   if(new_contact->id)
-   new_contact->id = num_objects.contacts;
-   new_contact->next = NULL;
-   if(new_contact->id>0)
      contact_list[new_contact->id - 1].next = new_contact;
  +
-   num_objects.contacts++;
  +
  return new_contact;

This is also crap. new_contact->id is an unsigned int, so if it is anything at all, it will be > 0. It's also just a stupid aesthetic change that has no impact on either logic or maintainability what so ever. Useless churn, in other words.

@@ -1200,10 +1212,5 @@ commandsmember *add_service_notification_command_to_contact(contact *cntct, char

```
/* duplicate vars */
```
-   if((new_commandsmember->command = (char *)strdup(command_name)) == NULL)

## \-       result = ERROR;
-   /\* handle errors */
-   if(result == ERROR) {
-       my_free(new_commandsmember->command);
-   if((new_commandsmember->command = (char *)strdup(command_name)) == NULL) {
      my_free(new_commandsmember);
      return NULL;

This actually makes sense (and yes, I realized I just sounded surprised), but not as part of the commit with the message you posted. Gather up similar patches (on top of current master, please) and commit all of them in a patch with a message like "cleanup: Remove superfluous variables

It's a good idea because...
"

Thanks.

**Comment by willixix** [comment]
That's why I said I fixed error handling on memory allocation, which did have a few bugs, but if your system is out of memory you're probably screwed anyway. Everything else is not important, however I personally think its better to be on the safe side and check for NULL (which is what the code been doing all over) and initialize to NULL as well, and I was trying to locate bug too when doing all these. free_object functions are not called often enough that you loose much ops by double-checking. As for ++ it makes for a more readable code.

Its in a separate commit and I think I can reverse the commit, but it seems to want to automatically add to pull request when I push and does not allow me to make a separate pull request. The only way to do that maybe to create a separate branch.

BTW, the current code is not stable. Previous code from 2 weeks ago was not either. The system stops checking after a few hours. Current code is even worse because plugins start to time out, previous just stopped doing checks. I was hoping latest code would be better and I'd not have to do debugging.

**Comment by ageric** [comment]
On 08/29/2012 09:35 PM, William Leibzon wrote:

> That's why I said I fixed error handling on memory allocation, which
> did have a few bugs, but if your system is out of memory you're
> probably screwed anyway. Everything else is not important, however I
> personally think its better to be on the safe side and check for NULL
> (which is what the code been doing all over) and initialize to NULL
> as well, and I was trying to locate bug too when doing all these.
> free_object functions are not called often enough that you loose much
> ops by double-checking. As for ++ it makes for a more readable code.
> 
> Its in a separate commit and I think I can reverse the commit, but it
> seems to want to automatically add to pull request when I push and
> does not allow me to make a separate pull request. The only way to do
> that maybe to create a separate branch.

The ++ and if(NULL) changes are in the same patch, which prevents me
from taking them. You could just revert that patch entirely and put
the strtok() patch on top of the modified tree so I can take that one
(assuming it's sane, ofcourse).

Btw, "git request-pull" lets you have more finegrained control over
the pull requests.

> BTW, the current code is not stable. Previous code from 2 weeks ago
> was not either. The system stops checking after a few hours. Current
> code is even worse because plugins start to time out, previous just
> stopped doing checks. I was hoping latest code would be better and
> I'd not have to do debugging.

I'm well aware of that. It's currently in alpha, as is pretty well
known. With 20k added lines and 30k removed ones I'm pretty sure
there are more of them lurking in the code.

**Comment by willixix** [comment]
I cleaned it. And my_free() macro was actually a problem, it did not work with loops which is another reason it was failing in that free contacts addresses inner loop. Like I said, it doesn't hurt to be more careful and double check on NULL.

I'll get strtok patch as a separate pull request from another branch in a week once I merge latest code and test it more. Its a lot of code changes there, more than here although mostly very repetitive.

**Comment by ageric** [comment]
On 08/30/2012 04:41 PM, William Leibzon wrote:

> I cleaned it. And my_free() macro was actually a problem, it did not
> work with loops which is another reason it was failing in that free
> contacts addresses inner loop.

How so?

#define my_free(ptr) do { if(ptr) { free(ptr); ptr = NULL; } } while(0)

Really shouldn't be able to screw up any kind of looping.

> Like I said, it doesn't hurt to be
> more careful and double check on NULL.

Double-checking is for humans who might fail the first check.
Getting a computer to do it is only a good idea if you worry about
heisenbugs.

> I'll get strtok patch as a separate pull request from another branch
> in a week once I merge latest code and test it more. Its a lot of
> code changes there, more than here although mostly very repetitive.

That would be good. Thanks.

File: base/config.c [hunk 1]
```diff
@@ -1549,8 +1553,8 @@ int pre_flight_check(void) {
```
Comment (ageric) [inline base/config.c:1552]: This chunk is unrelated.

File: base/config.c [hunk 2]
```diff
@@ -1107,7 +1107,7 @@ int read_main_config_file(char *main_config_file) {
 
 		else if(!strcmp(variable, "aggregate_status_updates")) {
 
-			/* DEPRECATED - ALL UPDATED ARE AGGREGATED AS OF NAGIOS 3.X */
```
Comment (ageric) [inline base/config.c:1110]: Unrelated.

File: common/objects.c [hunk 3]
```diff
@@ -221,25 +221,25 @@ timeperiod *add_timeperiod(char *name, char *alias) {
```
Comment (ageric) [inline common/objects.c:220]: Unrelated hunk

File: common/objects.c [hunk 4]
```diff
@@ -428,13 +428,13 @@ host *add_host(char *name, char *display_name, char *alias, char *address, char
```
Comment (ageric) [inline common/objects.c:427]: Unrelated hunk.

Besides... optimizing the error path which only leads to our early doom is more than a little useless.

File: common/objects.c [hunk 5]
```diff
@@ -460,138 +460,137 @@ host *add_host(char *name, char *display_name, char *alias, char *address, char
...
-	new_host->have_2d_coords = (have_2d_coords > 0) ? TRUE : FALSE;
-	new_host->x_3d = x_3d;
-	new_host->y_3d = y_3d;
-	new_host->z_3d = z_3d;
-	new_host->have_3d_coords = (have_3d_coords > 0) ? TRUE : FALSE;
-	new_host->should_be_drawn = (should_be_drawn > 0) ? TRUE : FALSE;
-	new_host->obsess_over_host = (obsess_over_host > 0) ? TRUE : FALSE;
-	new_host->retain_status_information = (retain_status_information > 0) ? TRUE : FALSE;
-	new_host->retain_nonstatus_information = (retain_nonstatus_information > 0) ? TRUE : FALSE;
+        if(result != ERROR) {
```
Comment (ageric) [inline common/objects.c:504]: Whitespace damage? Unrelated even so, so no thanks.

File: configure.in [hunk 6]
```diff
@@ -579,6 +579,11 @@ if test x$USE_EVENTBROKER = xyes; then
```
Comment (ageric) [inline configure.in:578]: This looks good. I'll take that.

File: lib/lnag-utils.h [hunk 7]
```diff
@@ -45,7 +45,7 @@
```
Comment (ageric) [inline lib/lnag-utils.h:44]: This is retarded, since it will no longer complain if you forget the semicolon.

You might as well get rid of the if() altogether, since every compiler in the world will remove it anyways.

Last, but definitely not least; With the if() it's impossible to use safely after 'else' without remembering to add the curly braces on all 'else' clauses.

This change is downright dangerous and could well lead to bugs that are extremely hard to find.

Consider this code:

if(x > 3)
  my_free(foo);
else
  do_something_else();

With your macro, the code will compile nicely, but the 'else' clause will belong to the if() statement in the macro, and so will never be executed.

File: base/config.c (hunk 2 above)
Comment (willixix) [inline base/config.c:1110]: Its a typo/spelling fix. Not even worth arguing about.

File: base/config.c (hunk 1 above)
Comment (willixix) [inline base/config.c:1552]: Reminder from when I was adding (void) to asprintf to avoid warnings. In this case it was easy to integrate checking its return value with nearby code. You should take it.

File: common/objects.c (hunk 3 above)
Comment (willixix) [inline common/objects.c:220]: cleans up error handing logic to exit immediately if possible

File: common/objects.c (hunk 4 above)
Comment (willixix) [inline common/objects.c:427]: No it is not. In the code everywhere it will continue to try to allocate memory even after it first failed and even though it knows all of it would have to be de-allocated because error already happened. I consider that to be badly written code and border-line bug.

File: common/objects.c (hunk 5 above)
Comment (willixix) [inline common/objects.c:504]: Nope, I purposely added tabs and brought this into inner loop of if OK. Same as with above, not doing anything extra if we know error condition already exists.

File: lib/lnag-utils.h (hunk 7 above)
Comment (willixix) [inline lib/lnag-utils.h:44]: I know compiler will optimize it out, in fact that's why I did it. And I knew of semicolon and decided its ok. But I did not think about the else case,  that is indeed dangerous and leads to hard to find bugs.

But I clearly saw that your existing do while did not work inside a for loop. I had to add "If not null" before my_free to make it work. It probably would have been enough to just  add { } for that loop, I can test that but it still means my_free is not safe how it is.

Its your call what to do but if you can't find a better replacement, would have to make a note to put my_free in {} in loops and go through code everywhere to make sure its there everywhere.

Comment (willixix) [inline lib/lnag-utils.h:44]: The else issue can be fixed by directly adding "else {}" to my macro. Still leaves semicolon, but I'd personally ignore that.

Comment (ageric) [inline lib/lnag-utils.h:44]: This is valid C (although not very sensible, but that's beside the point):
for (i = 0; i < 10; i++) do { i++; } while (i < 0);

So how could you clearly see that it doesn't work inside a for() loop?

And no, adding the 'else {}' to your macro _doesn't_ fix it, because if someone then uses the my_free macro and puts a semi-colon at the end of it (which is natural, because it looks like a function), then that semicolon will end the if() clause and leave the other else-clause dangling, which leads to a syntax error miscompile in code that looks just fine.

So..

if(x > 3)
  my_free(lolvar);
else
  do_stuff_with_lolvar();

would, with your version, expand to

if(x > 3)
  if(1) { if(lolvar) free(lolvar); lolvar = NULL; } else {};
else
  do_stuff_with_lolvar();

which is in no way legal C.

If you want your patches in the core, rebase them interactively onto some recent version of 'master' and remove all cruft so I can sanely review them without spending all day at it. Then I'll pick whatever patches measure up and drop the ones that don't.

**Comment by ageric** [comment]
I've commented enough. Most of the patches and changes in this pull request are horrible, and you don't even seem to see it (or your compiler is broken). I really like the savedata patch, but unless you clean it up considerably and interactive-rebase or cherry-pick it onto a new branch that doesn't have the codechurn patches and the horrible macro breakage stuff you've added, I'll drop that one too. My time is too limited and the patch doesn't add enough value.

In two weeks Nagios 4 goes into feature-freeze. If you want your patch in the core. I'd love to take the savedata patch before that. If it's not ready, it'll have to wait til Nagios 5, or get tacked on as a side-array so we can maintain the object api.
